targets/TARGET_NORDIC/TARGET_NRF5x/trng_api.c

```diff
@@ -0,0 +1,119 @@
+ * Copyright (c) 2017 Nordic Semiconductor ASA
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without modification,
+ * are permitted provided that the following conditions are met:
+ *
+ *   1. Redistributions of source code must retain the above copyright notice, this list
+ *      of conditions and the following disclaimer.
+ *
+ *   2. Redistributions in binary form, except as embedded into a Nordic Semiconductor ASA
+ *      integrated circuit in a product or a software update for such product, must reproduce
+ *      the above copyright notice, this list of conditions and the following disclaimer in
+ *      the documentation and/or other materials provided with the distribution.
+ *
+ *   3. Neither the name of Nordic Semiconductor ASA nor the names of its contributors may be
+ *      used to endorse or promote products derived from this software without specific prior
+ *      written permission.
+ *
+ *   4. This software, with or without modification, must only be used with a
+ *      Nordic Semiconductor ASA integrated circuit.
+ *
+ *   5. Any software provided in binary or object form under this license must not be reverse
+ *      engineered, decompiled, modified and/or disassembled.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#if defined(DEVICE_TRNG)
+
+#include "hal/trng_api.h"
+#include "hal/critical_section_api.h"
+
+#include "nrf_drv_rng.h"
+
+/* Keep track of instantiated FlashIAP objects. */
+static int nordic_trng_counter = 0;
+
+void trng_init(trng_t *obj)
+{
+    (void) obj;
+
+    /* Increment global counter inside critical section. */
+    hal_critical_section_enter();
+    nordic_trng_counter++;
+    int counter = nordic_trng_counter;
+    hal_critical_section_exit();
+
+    /* Initialize TRNG on first object only. */
+    if (counter == 1) {
+        nrf_drv_rng_init(NULL);
+    }
+}
+
+void trng_free(trng_t *obj)
+{
+    (void) obj;
+
+    /* Decrement global counter inside critical section. */
+    hal_critical_section_enter();
+    nordic_trng_counter--;
+    int counter = nordic_trng_counter;
+    hal_critical_section_exit();
+
+    /* Deinitialize TRNG when all objects have been freed. */
+    if (counter == 0) {
```

Comment:
This has a race: Say thread A calls `trng_free` and decreases `nordig_trng_counter` to `0`, but gets interrupted before the branch condition `counter == 0` is checked and `nrf_drv_rng_uninit()` is called, and say at that point another thread B calls `trng_init`, increasing `nordig_trng_counter` to `1` and calling `nrf_drv_rng_init`. Then, when thread A gets rescheduled, it calls `nrf_drv_rng_uninit`, leaving B's TRNG context unusable.

Reply:
Good point! Extending `hal_critical_section_exit` to after both `nrf_drv_rng_init` and `nrf_drv_rng_uninit` should be sufficient, right?

Reply:
@marcuschangarm Yes I think that should be fine.

Reply:
Actually we only need to synchronise ```nrf_drv_rng_uninit``` with ```nordig_trng_counter```. But it would be more robust to do both.